Add minimal Debian copyright file output #2235

pombredanne opened the pull request:
This is a minimal debian copyright file formatted output
It is dumb for now and creates one paragraph per file

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>
### Tasks

* [x] Reviewed [contribution guidelines](https://github.com/nexB/scancode-toolkit/blob/develop/CONTRIBUTING.rst)
* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [ ] Tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR
  Run [tests](https://scancode-toolkit.readthedocs.io/en/latest/contribute/contrib_dev.html#running-tests) locally to check for errors. 
* [x] Commits are in uniquely-named feature branch and has no merge conflicts 📁

pombredanne committed:
Add minimal Debian copyright file output #2235

This is a minimal debian copyright file formatted output
It is dumb for now and creates one paragraph per file
rather than one paragraph for a group of files

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

jelmer commented:
Thanks for adding this! Some initial observations (maybe for a follow-up DEP-5):

* This doesn't use the standardized DEP-5 terms for licenses - it should possibly have a dictionary mapping scancode license names to DEP-5 ? https://dep-team.pages.debian.net/deps/dep5/
* It doesn't recognise GPL-2.0+, instead saying "gpl-2.0 or gpl-3.0"
* it should exclude any full license files (currently includes e.g. COPYING)
* the actual license headers duplicate some lines from the original file

See e.g. https://pastebin.com/KLhCt1WD (generated by scancode) with https://salsa.debian.org/python-team/packages/dulwich/-/blob/5d9a44c4a56f01e2e9eed4a8f0e488bc7025ab93/debian/copyright (actual file)

pombredanne commented:
@jelmer thanks. 
so 1. https://salsa.debian.org/python-team/packages/dulwich/-/tree/5d9a44c4a56f01e2e9eed4a8f0e488bc7025ab93 is the exact tree you scanned
2. Let me attach the pastebin here
[dulwich.copyright.txt](https://github.com/nexB/scancode-toolkit/files/6057988/dulwich.copyright.txt)




>    This doesn't use the standardized DEP-5 terms for licenses - it should possibly have a dictionary mapping scancode license names to DEP-5 ? https://dep-team.pages.debian.net/deps/dep5/

Yes we need to draft a mapping. The question is also about all other licenses.
Say I have `GPL2+ WITH <some exception that ScanCode knows about>`
What should we dump then?
- use SPDX? : `GPL2+ WITH LicenseRef-scancode-exception-ScanCode-knows-about`
- use scancode : `GPL2+ WITH exception-ScanCode-knows-about`
What do we do about the exception-ScanCode-knows-about? a License paragraph?


>   It doesn't recognise GPL-2.0+, instead saying "gpl-2.0 or gpl-3.0"

GPL-2.0+ should be recognized OK, and I am not sure I saw exactly that in a dulwich scan rather mixing up things with some LGPL, and that's a bug. I looks like some dulwich notices are one of a kind, FWIW. That's still a bug on our side.

> it should exclude any full license files (currently includes e.g. COPYING)

We track the percentage of license text in a file and whether we matched against a full ext or not. Excluding full license files will need to be done in context of the license key (e.g. the mapping above). Here BTW, there is an important header in COPYING that states this is a choice of license that still may need to be reported?
So in any case we would need to have a way to exclude the standard full texts and only keep notices? 


>    the actual license headers duplicate some lines from the original file

Yes, that's another bug, likely in the `.dumps()` method of debian-inspector.

jelmer commented:
On Mon, Mar 01, 2021 at 11:08:35AM -0800, Philippe Ombredanne wrote:
> @jelmer thanks. 
> so 1. https://salsa.debian.org/python-team/packages/dulwich/-/tree/5d9a44c4a56f01e2e9eed4a8f0e488bc7025ab93 is the exact tree you scanned
> 2. Let me attach the pastebin here
> [dulwich.copyright.txt](https://github.com/nexB/scancode-toolkit/files/6057988/dulwich.copyright.txt)
> 
> >    This doesn't use the standardized DEP-5 terms for licenses - it should possibly have a dictionary mapping scancode license names to DEP-5 ? https://dep-team.pages.debian.net/deps/dep5/
> 
> Yes we need to draft a mapping. The question is also about all other licenses.
> Say I have `GPL2+ WITH <some exception that ScanCode knows about>`
> What should we dump then?
> - use SPDX? : `GPL2+ WITH LicenseRef-scancode-exception-ScanCode-knows-about`
> - use scancode : `GPL2+ WITH exception-ScanCode-knows-about`
> What do we do about the exception-ScanCode-knows-about? a License paragraph?

DEP-5 is explicit about this:

An exception or clarification to a license is signaled in plain text,
by appending with keywords exception to the short name. This document
provides a list of keywords that must be used when referring to the
most frequent exceptions. When exceptions other than these are in
effect that modify a common license by granting additional
permissions, you may use an arbitrary keyword not taken from the below
list of keywords. When a license differs from a common license because
of added restrictions rather than because of added permissions, a
distinct short name should be used instead of with keywords exception.

Only one exception may be specified for each license within a given
license specification. If more than one exception applies to a single
license, an arbitrary short name must be used instead.

There are only two reserved exceptions at the moment - OpenSSL and
Font. Other exceptions aren't defined - what you call them is up to
you, though it may be good to follow SPDX where possible?

> >   It doesn't recognise GPL-2.0+, instead saying "gpl-2.0 or gpl-3.0"
> GPL-2.0+ should be recognized OK, and I am not sure I saw exactly that in a dulwich scan rather mixing up things with some LGPL, and that's a bug. I looks like some dulwich notices are one of a kind, FWIW. That's still a bug on our side.
They are a bit different since it's dual-licensed under GPL2+ and
Apache-2. However, I would expect at most two licenses to show up
rather than four. It currently seems to match some lines multiple
times.

> > it should exclude any full license files (currently includes e.g. COPYING)

> We track the percentage of license text in a file and whether we matched against a full ext or not. Excluding full license files will need to be done in context of the license key (e.g. the mapping above). Here BTW, there is an important header in COPYING that states this is a choice of license that still may need to be reported?
> So in any case we would need to have a way to exclude the standard full texts and only keep notices? 
Yeah, that's what's done for debian/copyright files - lintian actually
warns when you reference full license files in debian/copyright.

Cheers,

Jelmer

-- 
Jelmer Vernooĳ <jelmer@jelmer.uk>
PGP Key: https://www.jelmer.uk/D729A457.asc

pombredanne committed:
Build copyright statements correctly

And not plain strings.
This is a fix related to https://github.com/nexB/debian-inspector/issues/10

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Use new debian-inspector

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Bump urllib3

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Do not crash if network op fails

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Align NOTICEs content

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Add notice to output

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>
